### PR TITLE
feat: Add chat with web data

### PR DESCRIPTION
**Pull request by leehuwuj**
Add a new option to chat with web data (only available in fastapi backend), use [WholeSite](https://llamahub.ai/l/web-whole_site) data loader from LlamaHub to crawl and load website data.

### CODE REVIEW

File: packages/create-llama/helpers/index.ts [hunk 1]
```diff
@@ -168,18 +179,23 @@ export const installTemplate = async (
...
-      props.packageManager,
-      props.engine,
-      props.openAiKey,
-      props.vectorDb,
-      props.contextFile,
-    );
+    if (props.engine === "context") {
+      if (props.dataSource === "file") {
+        // Copy test pdf file
+        // let dataSourceConfig = props.dataSourceConfig as FileSourceConfig;
```
Comment (marcusschiesser) [inline packages/create-llama/helpers/index.ts:189]: assert instead of comment? (or leave it)

File: packages/create-llama/helpers/types.ts [hunk 2]
```diff
@@ -6,6 +6,15 @@ export type TemplateEngine = "simple" | "context";
 export type TemplateUI = "html" | "shadcn";
 export type TemplateVectorDB = "none" | "mongo" | "pg";
 export type TemplatePostInstallAction = "none" | "dependencies" | "runApp";
+export type TemplateDataSource = "none" | "file" | "web";
```
Comment (marcusschiesser) [inline packages/create-llama/helpers/types.ts:9]: @leehuwuj good idea to introduce the datasource type! How about putting type and config in one object? (to simplify parameter passing)
```suggestion
export type TemplateDataSource = {
  type: "none" | "file" | "web";
  config: TemplateDataSourceConfig;
}  
```

File: packages/create-llama/templates/components/vectordbs/python/mongo/index.py [hunk 3]
```diff
@@ -20,4 +20,4 @@ def get_chat_engine():
     )
     index = VectorStoreIndex.from_vector_store(store, service_context)
     logger.info("Finished connecting to index from MongoDB.")
-    return index.as_chat_engine(similarity_top_k=5)
+    return index.as_chat_engine(chat_mode="context", similarity_top_k=5)
```
[the remark this answers is not in the document]
Comment (leehuwuj) [reply packages/create-llama/templates/components/vectordbs/python/mongo/index.py:23]: in my experience, chat_mode="context" gives better result than default one, especially we only apply theses change for vectordb code where user did choose to chat with context before.

File: packages/create-llama/helpers/types.ts (hunk 2 above)
Comment (leehuwuj) [reply packages/create-llama/helpers/types.ts:9]: great suggestion! I'll apply this one. Thank you!

File: packages/create-llama/helpers/types.ts [hunk 4]
```diff
@@ -15,8 +27,8 @@ export interface InstallTemplateArgs {
   template: TemplateType;
   framework: TemplateFramework;
   engine: TemplateEngine;
-  contextFile?: string;
   ui: TemplateUI;
+  dataSource: TemplateDataSource;
```
Comment (marcusschiesser) [inline packages/create-llama/helpers/types.ts:31]: @leehuwuj did you test this with llamapacks and community templates? pretty sure dataSource would then be undefined and your code might crash

File: packages/create-llama/templates/components/vectordbs/python/mongo/index.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply packages/create-llama/templates/components/vectordbs/python/mongo/index.py:23]: AFAIK the default chat mode is `BEST` - so if `BEST` is not the best, it seems like a llama_index issue that should not be handled here.

File: packages/create-llama/helpers/types.ts (hunk 4 above)
Comment (leehuwuj) [reply packages/create-llama/helpers/types.ts:31]: @marcusschiesser they are working well, but you are right, we may have undefined dataSource so let me change it to optional attribute

File: packages/create-llama/questions.ts [hunk 5]
```diff
@@ -389,6 +396,11 @@ export const askQuestions = async (
         },
         handlers,
       );
+      // Initialize with default config
+      program.dataSource = {
```
Comment (marcusschiesser) [inline packages/create-llama/questions.ts:400]: @leehuwuj this is sufficient as you already define the default above:
```suggestion
      // Initialize with default config
      program.dataSource = getPrefOrDefault("dataSource");
```

**Review by marcusschiesser — APPROVED** [review]

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**